### PR TITLE
Refactoring and extending article spam behavior

### LINKED ISSUE #15396: Expand spam trigger terms to look at entire post
**Is your feature request related to a problem? Please describe.**
As an admin battling spammers, I'd like to ensure that our spam triggers terms are as useful as possible.

**Describe the solution you'd like**
Ensure that our anti-spam keywords set in `Config > Rate limits and anti-spam > Spam trigger terms` looks at the entire post, including links.

**Additional context**
Will need to decide whether this looks at the body markdown or HTML.

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [X] Refactor
- [X] Bug Fix

## Description

Prior to this refactor, we were checking an Article's title for
spaminess.  This change now checks the Article's title and
body_markdown.

In addition, the PR encapsulates the regular expression implementations in
favor of asking the Settings::RateLimit class to determine if the passed
in text is "spammy".

Furthermore, instead of having lots of inline logic within the article,
this refactor introduces a `Spam::ArticleHandler` class.  This helps
tidy up the already busy Article specs by delegating the business logic
of Spam handling to it's own class.

There are further refactors for Comments and Users that would follow
this pattern, but this change is more important to get out the door.

## Related Tickets & Documents

Closes #15396

## QA Instructions, Screenshots, Recordings

Define a spam term, then create an article in which you add the spam term to the markdown.

### UI accessibility concerns?

None

## Added/updated tests?

- [X] Yes

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [X] I'm not sure how best to communicate this change and need help

### CODE REVIEW

**Review by Ridhwana — APPROVED** [review]
@jeremyf I reviewed the code and tested it through the console, and it works great! Pulling the code in to it's own service is a really nice refactor!🚀

Something food for thought✨  - I think it would be useful to give more detailed QA instructions and outline the expected outcomes for the QA. I knew my way around the admin interface to add spam terms + check reaction etc. and I could do it through the rails console, but in general it may help in the future to have slightly more detailed instructions so that anyone can review your PR with more ease.

**Comment by jeremyf** [comment]
> Something food for thought✨ - I think it would be useful to give more detailed QA instructions and outline the expected outcomes for the QA. I knew my way around the admin interface to add spam terms + check reaction etc. and I could do it through the rails console, but in general it may help in the future to have slightly more detailed instructions so that anyone can review your PR with more ease.

I have another PR that will build on this, I'll include QA steps in that next one.  

Question: my strong preference is to _not_ require someone to go through the UI for this refactor.  Would it be adequate to give `rails console` guidance instead?

**Comment by msarit** [comment]
> Question: my strong preference is to not require someone to go through the UI for this refactor. Would it be adequate to give rails console guidance instead?

I think giving both is good. I have had PRs where I listed QA steps for both the UI and the console.

**Review by msarit — COMMENTED** [review]
I tested this PR in the UI, and I am not getting the behavior that I expected. 😟 

Now it may be that my expectations are off. Here is a video showing what I experienced:

https://www.loom.com/share/56c744797272427e94b3c79b88b06294

Let me know if my expectation is off, or my testing approach is off as well 🙏🏾

**Comment by jeremyf** [comment]
@msarit Thanks for the video, it definitely helps to understand your expectations.

The present behavior (prior to this PR) is we allow posts with spam content, but perform some administrative tasks that escalate towards suspending the user.  This PR refactors by encapsulating the spam handling logic and also now checking title and body markdown.

**Comment by Ridhwana** [comment]
> Question: my strong preference is to _not_ require someone to go through the UI for this refactor. Would it be adequate to give `rails console` guidance instead?

I agree with Arit, I try and give guidance on both where applicable. Some reviewers prefer to test through the UI whereas others are to test through the console.

**Comment by jeremyf** [comment]
In this case, the test through the UI doesn't say all that much because your first offense is "free" (except that the mascot "vomit"-ed on your article).  The next offense also induces a "vomit" and a suspension.

So testing that requires a LOT of QA overhead, on what is a very minor adjustment (e.g. extracting a class and scanning an additional attribute)

**Comment by Ridhwana** [comment]
> In this case, the test through the UI doesn't say all that much because your first offense is "free" (except that the mascot "vomit"-ed on your article). The next offense also induces a "vomit" and a suspension.
> 
> So testing that requires a LOT of QA overhead, on what is a very minor adjustment (e.g. extracting a class and scanning an additional attribute).
> 
> All of which is to say, I'm super mindful of our bandwidth as people, and trying to find some point where we can optimize our energy expenditure.

I have a different view on this - I feel like providing more in depth QA guidance will actually optimize for energy expenditure in this case. When reviewing a PR, if I have to think about how to QA it as well figure out how to test there are no regressions on a refactor, I find myself spending more time on the review than I otherwise would have needed to.

In this case, to ensure that we didn't regress any functionality when doing the refactor, because we did move some code around, and add some new functionality,  I looked through the `Reactions` and made sure that the `vomits ` appeared when the spam word was in the title or body of the article. I also double checked the user's role and ensured that they are suspended on the third count.

**Comment by jeremyf** [comment]
> In this case, to ensure that we didn't regress any functionality when doing the refactor (because we moved some code around + add some new functionality) I looked through the `Reactions` and made sure that the `vomits ` appeared when the spam word was in the title or body of the article. I also double checked the user's role and ensured that they are suspended on the third count.

I did write specs to assert all of those pathways, so I see those as providing a backstop for regression.  I do also acknowledge that this conversation around where and what to test is the inevitable song and dance of the QA process, unit testing versus functional testing versus integration testing.

File: spec/models/article_spec.rb [hunk 1]
```diff
@@ -1033,56 +1033,10 @@ def build_and_validate_article(*args)
...
-        third_article.save
-        expect(article.user.suspended?).to be true
-        expect(Note.last.reason).to eq "automatic_suspend"
-      end
-
-      it "does not create vomit reaction if does not have matching title" do
-        article.save
-        expect(Reaction.last).to be nil
-      end
+      it "delegates spam handling to Spam::ArticleHandler" do
```
Comment (maestromac) [inline spec/models/article_spec.rb:1036]: Is this spec missing an expectation?

**Review by maestromac — APPROVED** [review]
Awesome refactor and enhancement! aside from my other comment, LGTM!

**Pull request merged**